src/compiler/checker.ts

```diff
@@ -27968,7 +27938,7 @@ namespace ts {
          * Return true if the given type is considered to have numeric property names.
          */
         function hasNumericPropertyNames(type: Type) {
-            return getIndexTypeOfType(type, IndexKind.Number) && !getIndexTypeOfType(type, IndexKind.String);
+            return getIndexInfosOfType(type).length === 1 && !!getIndexInfoOfType(type, numberType);

```

Comment:
This diff surprised me; I'd like to understand it.

Reply:
Yeah, it's a bit odd. For an element access expression `a[i]`, when `a` has _only_ a numeric index signature and `i` is the iteration variable of a `for (let i in a)`, we consider `i` to have type `number` in the element access expression. Basically, this is what enables `for (let i in a)` to work with arrays. Now that we can have any number of index signatures, it's not enough to just check we don't have a string index signature. Hence the change.

Reply:
Ah my confusion came from elsewhere: the diff highlighting made me misread it as `!` instead of `!!`

‼